internal/postrs/proof.go

```diff
@@ -44,10 +46,11 @@ func GenerateProof(dataDir string, challenge []byte, cfg config.Config, logger *
 	defer C.free(challengePtr)
 
 	config := C.Config{
-		k1:                C.uint32_t(cfg.K1),
-		k2:                C.uint32_t(cfg.K2),
-		k2_pow_difficulty: C.uint64_t(cfg.K2PowDifficulty),
-		pow_scrypt:        translateScryptParams(powScrypt),
+		k1: C.uint32_t(K1),
+		k2: C.uint32_t(K2),
+	}
+	for i, b := range powDifficulty {
+		config.pow_difficulty[i] = C.uchar(b)
 	}
```

Comment:
Why not do the same for the difficulty as for the challenge?

```golang
	diffPtr := C.CBytes(powDifficulty)
	defer C.free(diffPtr)

	config := C.Config{
		k1: C.uint32_t(K1),
		k2: C.uint32_t(K2),
		pow_difficulty: (*C.uchar)(diffPtr),
	}
```

avoids copying the data, but requires `pow_difficulty` to be a pointer instead of an array.

Reply:
It doesn't avoid copying the data because `C.CBytes(powDifficulty)` allocates memory and makes a copy. It will only add a layer of indirection and lose the benefit of having the length embedded in the type (`[32]byte` vs `[]byte`), which makes the API easier to use properly.

Reply:
`[32]byte` is a value type so it is copied when the underlying C.generate_proof function is called, while `[]byte` is a pointer and the data isn't copied.

I don't think it makes much difference between arrays and slices in terms of what is easier to use. We even use `[]byte` over `[x]byte` for most other types here, passing fewer bytes to the function also just fils them with zeros instead of causing an error during compilation: https://go.dev/play/p/i_5jDu8yc50

Reply:
>[32]byte is a value type so it is copied when C.generate_proof function is called a bit further down, while C.CBytes returns a pointer to the data that doesn't need to be copied again.

That's the same amount of copying, except `C.CBytes()` allocates on a heap (and copies the contents into there) while the array is kept on a stack. Anyhow, it's supposed to be called once every epoch.

>I don't think it makes much difference between arrays and slices in terms of what is easier to use.

I mean that an array, which is fixed length better expresses the API requirements.
- `[32]byte` says: "Pass me **exactly 32** bytes." 
- `[]byte` says: "Pass me **some** bytes

In other words, when I see a function signature:
```go
func Foo(val [32]byte)
```
- I immediately know that I must pass 32 bytes,
- I don't need to validate length of passed data (which might be necessary if `[]byte` is passed).

Reply:
> That's the same amount of copying, except C.CBytes() allocates on a heap (and copies the contents into there) while the array is kept on a stack.

Array does 2 copies: here in this `for` loop and again when calling `C.generate_proof` by passing the array via value to the function.

Slice only once: when calling `C.CBytes`, after that only the pointer to the array is passed to `C.generate_proof`.

Reply:
But you are right it's not a necessary optimization because generating a proof is only done every 2 weeks and compared to the rest of the process its an negligible improvement.